File: mobile/lib/models/participation.dart
import 'user.dart';
import 'game.dart';
import 'prediction.dart';

class Participation {
  final String id;
  final User user;
  final Game game;
  final List<Prediction> predictions;
  final int livesRemaining;
  final int totalPoints;
  final bool isEliminated;
  final int? eliminationWeek;
  final DateTime joinedAt;
  final DateTime lastActivityAt;
  final DateTime createdAt;
  final DateTime updatedAt;

  Participation({
    required this.id,
    required this.user,
    required this.game,
    required this.predictions,
    required this.livesRemaining,
    required this.totalPoints,
    required this.isEliminated,
    this.eliminationWeek,
    required this.joinedAt,
    required this.lastActivityAt,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Participation.fromJson(Map<String, dynamic> json) {
    return Participation(
      id: json['_id'] ?? json['id'],
      user: User.fromJson(json['userId'] is Map ? json['userId'] : {}),
      game: Game.fromJson(json['gameId'] is Map ? json['gameId'] : {}),
      predictions: (json['predictions'] as List<dynamic>?)
          ?.map((prediction) => Prediction.fromJson(prediction))
          .toList() ?? [],
      livesRemaining: json['livesRemaining'] ?? 3,
      totalPoints: json['totalPoints'] ?? 0,
      isEliminated: json['isEliminated'] ?? false,
      eliminationWeek: json['eliminationWeek'],
      joinedAt: json['joinedAt'] != null ? DateTime.parse(json['joinedAt']) : DateTime.now(),
      lastActivityAt: json['lastActivityAt'] != null ? DateTime.parse(json['lastActivityAt']) : DateTime.now(),
      createdAt: json['createdAt'] != null ? DateTime.parse(json['createdAt']) : DateTime.now(),
      updatedAt: json['updatedAt'] != null ? DateTime.parse(json['updatedAt']) : DateTime.now(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'user': user.toJson(),
      'game': game.toJson(),
      'predictions': predictions.map((prediction) => prediction.toJson()).toList(),
      'livesRemaining': livesRemaining,
      'totalPoints': totalPoints,
      'isEliminated': isEliminated,
      'eliminationWeek': eliminationWeek,
      'joinedAt': joinedAt.toIso8601String(),
      'lastActivityAt': lastActivityAt.toIso8601String(),
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }

  List<Prediction> get currentWeekPredictions {
    return predictions.where((p) => p.week == game.currentWeek).toList();
  }

  bool hasPredictionForMatch(String matchId, {int? week}) {
    if (week != null) {
      return predictions.any((p) => p.matchId == matchId && p.week == week);
    }
    return predictions.any((p) => p.matchId == matchId);
  }
}

File: backend/models/Participation.ts
import mongoose, { Document, Schema } from 'mongoose';

export interface IPrediction {
  week: number;
  matchId: string;
  selectedTeam: 'home' | 'visitor' | 'draw';
  isCorrect?: boolean;
  pointsEarned?: number;
  createdAt: Date;
}

export interface IParticipation extends Document {
  userId: mongoose.Types.ObjectId;
  gameId: mongoose.Types.ObjectId;
  predictions: IPrediction[];
  livesRemaining: number;
  totalPoints: number;
  isEliminated: boolean;
  eliminationWeek?: number;
  joinedAt: Date;
  lastActivityAt: Date;
  createdAt: Date;
  updatedAt: Date;
}

const PredictionSchema = new Schema({
  week: { type: Number, required: true, min: 1 },
  matchId: { type: String, required: true },
  selectedTeam: { 
    type: String, 
    required: true, 
    enum: ['home', 'visitor', 'draw'] 
  },
  isCorrect: { type: Boolean },
  pointsEarned: { type: Number, default: 0 },
  createdAt: { type: Date, default: Date.now }
});

const ParticipationSchema: Schema = new mongoose.Schema({
  userId: { 
    type: Schema.Types.ObjectId, 
    ref: 'User', 
    required: true 
  },
  gameId: { 
    type: Schema.Types.ObjectId, 
    ref: 'Game', 
    required: true 
  },
  predictions: [PredictionSchema],
  livesRemaining: { 
    type: Number, 
    required: true, 
    min: 0 
  },
  totalPoints: { 
    type: Number, 
    default: 0,
    min: 0
  },
  isEliminated: { 
    type: Boolean, 
    default: false 
  },
  eliminationWeek: { 
    type: Number 
  },
  joinedAt: { 
    type: Date, 
    default: Date.now 
  },
  lastActivityAt: { 
    type: Date, 
    default: Date.now 
  }
}, {
  timestamps: true
});

// Indexes for better performance
ParticipationSchema.index({ userId: 1, gameId: 1 }, { unique: true });
ParticipationSchema.index({ gameId: 1, isEliminated: 1 });
ParticipationSchema.index({ userId: 1, isEliminated: 1 });

export default mongoose.model<IParticipation>('Participation', ParticipationSchema);
